internal/services/authorization/role_definition_data_source.go

```diff
@@ -4,178 +4,245 @@
+				scope = commonids.NewSubscriptionID(metadata.Client.Account.SubscriptionId).ID()
 			}
-			if roleDefinitions.Values()[0].ID == nil {
-				return pluginsdk.NonRetryableError(fmt.Errorf("loading Role Definition List: values[0].ID is nil '%s'", name))
+
+			// search by name
+			var id roledefinitions.ScopedRoleDefinitionId
+			var role roledefinitions.RoleDefinition
+			if config.Name != "" {
+				// Accounting for eventual consistency
+				deadline, ok := ctx.Deadline()
+				if !ok {
+					return fmt.Errorf("internal error: context had no deadline")
+				}
+				err := pluginsdk.Retry(time.Until(deadline), func() *pluginsdk.RetryError {
+					roleDefinitions, err := client.List(ctx, commonids.NewScopeID(scope), roledefinitions.ListOperationOptions{
+						Filter: pointer.To(fmt.Sprintf("roleName eq '%s'", config.Name)),
+					})
+					if err != nil {
+						return pluginsdk.NonRetryableError(fmt.Errorf("loading Role Definition List: %+v", err))
+					}
+					if roleDefinitions.Model == nil {
+						return pluginsdk.RetryableError(fmt.Errorf("loading Role Definition List: `Model` was nil"))
+					}
+					if len(*roleDefinitions.Model) != 1 {
+						return pluginsdk.RetryableError(fmt.Errorf("loading Role Definition List: could not find role '%s'", config.Name))
+					}
+					if (*roleDefinitions.Model)[0].Id == nil {
+						return pluginsdk.NonRetryableError(fmt.Errorf("loading Role Definition List: values[0].ID is nil '%s'", config.Name))
+					}
+
+					id = roledefinitions.NewScopedRoleDefinitionID(scope, *(*roleDefinitions.Model)[0].Name)
+					return nil
+				})
+				if err != nil {
+					return err
+				}
+			} else {
+				id = roledefinitions.NewScopedRoleDefinitionID(scope, defId)
 			}
 
-			defId = *roleDefinitions.Values()[0].ID
-			role, err = client.GetByID(ctx, defId)
+			resp, err := client.Get(ctx, id)
 			if err != nil {
-				return pluginsdk.NonRetryableError(fmt.Errorf("getting Role Definition by ID %s: %+v", defId, err))
+				return fmt.Errorf("retirving %s: %+v", id, err)
 			}
-			return nil
-		})
-		if err != nil {
-			return err
-		}
-	} else {
-		var err error
-		role, err = client.Get(ctx, scope, defId)
-		if err != nil {
-			return fmt.Errorf("loading Role Definition: %+v", err)
-		}
+
+			if resp.Model == nil {
+				return fmt.Errorf("retirving %s: `Model` was nil", id)
+			}
+
+			role = *resp.Model
+
+			state := ArmRoleDefinitionDataSourceModel{
+				Scope:            id.Scope,
+				RoleDefinitionId: id.RoleDefinitionId,
+			}
+
+			if props := role.Properties; props != nil {
+				state.Name = pointer.From(props.RoleName)
+				state.Type = pointer.From(props.Type)
+				state.Description = pointer.From(props.Description)
+				state.Permissions = flattenDataSourceRoleDefinitionPermissions(props.Permissions)
+				state.AssignableScopes = pointer.From(props.AssignableScopes)
+			}
+
+			// for data source, we just use the resource id without scope instead of scoped id.
+			// to keep consistent with old versions.
```

Comment:
Is this the same as the original role definiton id? I assume the original is like `/subscriptions/sub1/providers/Microsoft.Authorization/roleDefinitions/def1`, but here it will be `/providers/Microsoft.Authorization/roleDefinitions/def1`?

Whilst, I believe the resource id of a data source isn't a big deal, which means we can consider changing it to simply `id.ID()`?